core/services/workflows/syncer/workflow_registry.go

```diff
@@ -353,13 +402,192 @@ func (w *workflowRegistry) readRegistryEvents(ctx context.Context, don capabilit
+// Differences are handled by the event handler by creating events that are sent to the events channel for handling.
+func (w *workflowRegistry) workflowMetadataToEvents(ctx context.Context, workflowMetadata []GetWorkflowMetadata, donID uint32) {
+	var events []workflowAsEvent
+
+	// Keep track of which of the engines in the engineRegistry have been touched
+	engineKeyMap := map[string]bool{}
+
+	for _, wfMeta := range workflowMetadata {
+		// TODO: ensure that the WorkflowRegisteredEvent clears the engine registry as the very last step
+		engine, getErr := w.engineRegistry.Get(EngineRegistryKey{Owner: wfMeta.Owner, Name: wfMeta.WorkflowName})
+		currWfID := hex.EncodeToString(wfMeta.WorkflowID[:])
+
+		// if there is no engine in the engine registry
+		// then handle as registered event
+		if getErr != nil {
+			toRegisteredEvent := WorkflowRegistryWorkflowRegisteredV1{
+				WorkflowID:    wfMeta.WorkflowID,
+				WorkflowOwner: wfMeta.Owner,
+				DonID:         wfMeta.DonID,
+				Status:        wfMeta.Status,
+				WorkflowName:  wfMeta.WorkflowName,
+				BinaryURL:     wfMeta.BinaryURL,
+				ConfigURL:     wfMeta.ConfigURL,
+				SecretsURL:    wfMeta.SecretsURL,
+			}
+			events = append(events, workflowAsEvent{
+				Data:      toRegisteredEvent,
+				EventType: WorkflowRegisteredEvent,
+			})
+			engineKeyMap[currWfID] = true
+			continue
+		}
+
+		// if the workflow engine is in the registry, but the metadata has changed
+		// then handle as updated event
+		prevWfID := hex.EncodeToString(engine.workflowID[:])
+		if currWfID != prevWfID {
+			toUpdatedEvent := WorkflowRegistryWorkflowUpdatedV1{
+				OldWorkflowID: engine.workflowID,
+				NewWorkflowID: wfMeta.WorkflowID,
+				WorkflowOwner: wfMeta.Owner,
+				DonID:         wfMeta.DonID,
+				WorkflowName:  wfMeta.WorkflowName,
+				BinaryURL:     wfMeta.BinaryURL,
+				ConfigURL:     wfMeta.ConfigURL,
+				SecretsURL:    wfMeta.SecretsURL,
+			}
+			events = append(events, workflowAsEvent{
+				Data:      toUpdatedEvent,
+				EventType: WorkflowUpdatedEvent,
+			})
+			engineKeyMap[currWfID] = true
+			continue
+		}
+
+		// if the workflow engine is in the registry, the metadata is the same, and the status is active
+		// then the workflow is unchanged. mark as seen.
+		if WorkflowStatus(wfMeta.Status) == WorkflowStatusActive {
+			engineKeyMap[wfMeta.WorkflowName+hex.EncodeToString(wfMeta.Owner)] = true
+			continue
+		}
+
+		// if the workflow engine is in the registry, the metadata is the same, but the status isn't active
+		// then the workflow has been paused. NOOP - to be handled below as a deleted event, which clears the DB workflow spec.
+	}
+
+	// Shut down engines that are no longer in the contract's latest workflow metadata state
+	allEngines := w.engineRegistry.GetAll()
+	for _, engine := range allEngines {
+		prevWfID := hex.EncodeToString(engine.workflowID[:])
+		_, exists := engineKeyMap[prevWfID]
```

Comment:
should this not be wfowner+wfname?

Reply:
Amended. My first intuition and what you said are right. It has to be wfowner+wfname since the handler for update currently does the delete.